trecs/components/users.py

```diff
@@ -347,3 +359,185 @@ def update_profiles(self, item_attributes):
+            :math:`|U|\\times\\text{num_items_per_iter}` matrix with
+            recommendations and new items.
+
+        item_attributes: :obj:`numpy.ndarray`):
+            A :math:`|A|\\times|I|` matrix with item attributes.
+
+        Returns
+        ---------
+            Array of interactions s.t. element :math:`interactions_{u(t)}` represents the
+            index of the item selected by user `u` at time `t`. Shape: |U|
+
+        Raises
+        -------
+
+        ValueError
+            If :attr:`interact_with_items` is None and there is not `item`
+            parameter.
+        """
+        if self.interact_with_items is not None:
+            return self.interact_with_items(*args, **kwargs)
+        items_shown = kwargs.pop("items_shown", None)
+        item_attributes = kwargs.pop("item_attributes", None)
+        if items_shown is None:
+            raise ValueError("Items can't be None")
+        reshaped_user_vector = self.user_vector.reshape((items_shown.shape[0], 1))
+        interaction_scores = self.actual_user_scores[reshaped_user_vector, items_shown]
+
+        self.log("User scores for given items are:\n" + str(interaction_scores))
+        item_utilities = self.dn_utilities(interaction_scores)
+        sorted_user_preferences = item_utilities.argsort()[:, -1]
+        interactions = items_shown[self.user_vector, sorted_user_preferences]
+        self.log("Users interact with the following items respectively:\n" + str(interactions))
+
+        if self.drift > 0:
+            if item_attributes is None:
+                raise ValueError("Item attributes can't be None if user preferences are dynamic")
+            # update user profiles based on the attributes of items they
+            # interacted with
+            interact_attrs = item_attributes.T[interactions, :]
+            self.update_profiles(interact_attrs)
+            # update user scores
+            self.compute_user_scores(item_attributes)
+        return interactions
+
+    def normalize_values(self, user_item_scores):
+        """
+        Calculating the expression for :math:`z(\\textbf{v})` in the equation
+        :math:`z(\\textbf{v})+\\mathbf{\\eta}`.
+
+        Parameters
+        -----------
+
+        user_item_scores: :obj:`array_like`
+            The element at index :math:`i,j` should represent user :math:`i`'s
+            context-independent value for item :math:`j`.
+            Dimension: :math:`|U|\\times|I|`
+
+        Returns
+        --------
+            normed_values: :obj:`numpy.ndarray`
+                Probabilities of a user making a particular choice. All probabilities
+                for a given row will sum up to 1.
+        """
+        summed_norms = np.linalg.norm(user_item_scores, ord=self.beta, axis=1)
+        denom = self.sigma + np.multiply(self.omega, summed_norms)
+        return np.divide(user_item_scores.T, denom)  # now |I| x |U|
+
+    def dn_utilities(self, user_item_scores):
```

Comment:
I would like all function names to start with a verb. I believe this is not 100% enforced, but this is something I've been meaning to fix. How about something like `calculate_dn_utilities`? I don't have strong preferences, so as long as it's a reasonable verb I'm fine with it

Reply:
Sounds good to me - done and altered to `calc_dn_utilities` for brevity!